enhancements/cluster-logging/forwarder-tagging.md

```diff
@@ -0,0 +1,139 @@
+---
+title: Forwarder adds Tags to Outbound Messages
+authors:
+  - "@alanconway"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-01
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Forwarder adds Tags to Outbound Messages
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is implementable
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Add configuration so the Forwarder can attach name+value fields to each message that is forwarded to a remote destination.
+For example: add a `datacenter-id` field with a value identifying a datacenter to every log message forwarded.
+
+Tags are associated with a forwarder `pipeline`. Different tags can be set based on `input` filters and `output` destinations. 
+For example `audit` logs can be tagged differently from `application` logs.
+
+This is a cluster-wide control point for _cluster administrators_, independent of the content of logs.
+These tags cannot be modified or deleted by applications in the local cluster.
+
+## Motivation
+
+### Goals
+
+* Cluster admin can tag outgoing messages at the forwarder.
+* Set fixed tags based on `output` destination.
+* Set fixed tags based on `input` origin, allow use of all present and future `input` filtering features.
+
+### Non-Goals
+
+* Tags are set at configuration, no computed tags based on message content.
+* Does not replace or interfere with other tagging mechanisms prior to the forwarder.
+
+## Proposal
+
+Add a `tags` map to the `pipeline` spec. See [User Stories](#user-stories) for examples.
+The tags are applied to all logs passing through that pipeline.
+
+### User Stories
+
+#### As cluster admin I want to label all outgoing messages with a cluster ID
+
+```
+pipeline:
+  tags: { cluster-id: C1234 }
+  inputRefs: [application, infrastructure, audit]
+  outputRefs: [somewhere]
+```
+
+#### As cluster admin I want to label different message types differently.
+
+```
+pipeline:
+  tags: { logType: normal }
+  inputRefs: [application]
+  outputRefs: [somewhere]
+pipeline:
+  tags: { logType: special }
+  inputRefs: [infrastructure]
+  outputRefs: [somewhere]
+pipeline:
+  tags: { logType: secret }
+  inputRefs: [audit]
+  outputRefs: [somewhere]
+```
+
+### Risks and Mitigations
+
+Risk: Excessive use of tags may bloat outgoing log traffic.
+
+Mitigation: Cluster admins control use, with great power comes great responsibility.
+
+## Design Details
+
+Add a `tags` map to the `pipeline` spec, each name-value pair from the tags map
+is added as a tag to each log message. Forwarder includes tag directives in
+`fluentd` config.
+
+### Tag format
+TODO: define where tags are placed in message. May need further API options to control?
```

Comment:
Maybe we should take a hint from Kubernetes here and be consistent with their terminology and purpose: 'labels' or 'static_labels' or 'additional_labels'.  Personally I prefer they just be 'labels' and we merge them with other labels we find.  Labels are supposed to be the way user's slice and dice their applications and are used for query.  Imagine doing a query where i want to filter on both those defined in the deployment and the static one available.  Wouldn't it be easier to just use the same key?

Reply:
I followed the kubernetes example in the normalized record: openshift.labels. Its consistent and leave a place to put other stuff in future.